modules/service/src/main/scala/lucuma/odb/service/GuideEnvironmentService.scala

```diff
@@ -0,0 +1,469 @@
+        val request                        = Request[F](Method.GET,
+                                 CatalogSearch.gaiaSearchUri(query),
+                                 headers = Headers(("x-requested-with", "XMLHttpRequest"))
+        )
+        httpClient
+          .stream(request)
+          .flatMap(
+            _.body
+              .through(utf8.decode)
+              .through(CatalogSearch.guideStars[F](CatalogAdapter.Gaia3Lite))
+          )
+          .compile
+          .toList
+          .map(_.collect { case Right(s) => GuideStarCandidate.siderealTarget.get(s).at(obsTime) })
+          .map(
+            NonEmptyList
+              .fromList(_)
+              .toRight(
+                Error
+                  .GaiaError(s"No valid guide star candidates were returned for observation $oid.")
+              )
+          )
+          .handleError(e => Error.GaiaError(e.getMessage()).asLeft)
+      }
+
+      // TODO: use caching
+      def getCandidates(
+        oid:      Observation.Id,
+        obsTime:  Instant,
+        tracking: ObjectTracking
+      ): F[Either[Error, NonEmptyList[GuideStarCandidate]]] =
+        (for {
+          query      <- EitherT.fromEither(getGaiaQuery(oid, obsTime, tracking, probeArm.candidatesArea))
+          candidates <- EitherT(callGaia(oid, obsTime, query))
+        } yield candidates).value
+
+      extension [D](steps: NonEmptyList[Step[D]])
+        def offsets: List[Offset] = steps.collect { case Step(_, _, StepConfig.Science(offset, _), _, _, _) =>
+          offset
+        }
+
+      def offsetsFromExecConfig[S, D](
+        oid:        Observation.Id,
+        execConfig: ExecutionConfig[S, D]
+      ): Option[NonEmptyList[Offset]] =
+        execConfig.science
+          .map { s =>
+            s.nextAtom.steps.offsets ++
+              s.possibleFuture.flatMap(_.steps.offsets)
+          }
+          .flatMap(NonEmptyList.fromList)
+
+      def getOffsets(
+        pid: Program.Id,
+        oid: Observation.Id
+      ): F[Either[Error, Option[NonEmptyList[Offset]]]] =
+        generator(commitHash, itcClient, plannedTimeCalculator)
+          .digest(pid, oid)
+          .map {
+            _.leftMap(Error.GeneratorError(_))
+              .map { d => NonEmptyList.fromFoldable(d.science.offsets) }
+          }
+
+      def getPositions(
+        oid:                Observation.Id,
+        site:               Site,
+        posAngleConstraint: PosAngleConstraint,
+        offsets:            Option[NonEmptyList[Offset]],
+        tracking:           ObjectTracking,
+        obsTime:            Instant
+      ): Either[Error, NonEmptyList[AgsPosition]] = {
+        val angles     =
+          posAngleConstraint.anglesToTestAt(site, tracking, obsTime).map(_.sorted(Angle.AngleOrder))
+        val newOffsets = offsets match
+          case None          => NonEmptyList.of(Offset.Zero)
+          case Some(offsets) => offsets.prepend(Offset.Zero).distinct
```

Comment:
If there are offsets, why do we need to include (0, 0)?

Reply:
I think I need to defer to @cquiroz again. More plagiarism...

Reply:
I noted that when we get offsets we don't always get the position at zero, this `prepend` ensures that is always present

Reply:
I think if there are offsets but none of them are at (0, 0), then there should be no datasets collected at (0, 0) in which case the guide star doesn't need to be valid there.

Reply:
So, should I remove the prepend? @cquiroz @swalker2m

Reply:
I'm trying to remember why.

We are finding the region that intersect all offsets such that a GS is valid at all of them. But that would be valid even if there is no (0, 0) and the list of offsets is certainly non empty

I'd say let's remove it here and in explore 

Also we should consider moving some of this logic to `lucuma-ags` such that it is common for the odb and explore

Reply:
OK. I'll remove it here and put in a PR for explore.
> Also we should consider moving some of this logic to `lucuma-ags` such that it is common for the odb and explore

Yes. I can examine that as part of creating a gaia client.